Add FXIOS-5790 [v112] rust sync manager added

lougeniaC64 opened the pull request:
This is the third task in #13278 which is adding the `RustSyncManager` class for the sync manager integration. This class will replace the existing `BrowserSyncManager` class.

**Note:** The changes in `RustRemoteTabs.swift`, `RustSyncManagerAPI.swift`, and `SyncTelemetryUtils.swift` should have been included in #13330 but were omitted because of the Glean build issues.

lougeniaC64 commented:
@nbhasin2 @mhammond @tarikeshaq Ready for review, and I'm open to any all feedback. This is the bulk of the sync manager integration logic, and the replacement for the current [BrowserSyncManager](https://github.com/mozilla-mobile/firefox-ios/blob/main/Providers/Profile.swift#L751) class, so this should be logic that both of our teams can easily grok and feel comfortable maintaining.

nbhasin2 commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,635 @@
+    // But it's possible that we'll finish a side-effect sync after we've ditched the
+    // profile as a whole, so we hold on to our Prefs, potentially for a little while
+    // longer. This is safe as a strong reference, because there's no cycle.
+    unowned fileprivate let profile: BrowserProfile
```
```suggestion
    private let profile: BrowserProfile
```

Is there a reason you need `unowned` here? We use that mostly in closures inside the capture list `[unowned let...] in` but I am not sure what benefit we get for using unowned here. It might end up being dangerous because its a strong reference which has no relation to retain and release cycle. So you can see random crashes.

lougeniaC64 replied:
No other reason that it was in [BrowserSyncManager](https://github.com/mozilla-mobile/firefox-ios/blob/7c90d922f0b2cdad3d311f6649ecc2c04d4e20f6/Providers/Profile.swift#L757).

nbhasin2 replied:
The interesting fact is that its an 8-year-old code that still works!

Personally, I avoid the `unowned` unless I am very confident in the lifecycle of the variable which might have been the case 8 years ago.

nbhasin2 commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,635 @@
+    // profile as a whole, so we hold on to our Prefs, potentially for a little while
+    // longer. This is safe as a strong reference, because there's no cycle.
+    unowned fileprivate let profile: BrowserProfile
+    fileprivate let prefs: Prefs
+    fileprivate var constellationStateUpdate: Any?
+    fileprivate var syncTimer: Timer?
+    fileprivate var backgrounded: Bool = true
```
```suggestion
    private let prefs: Prefs
    private var constellationStateUpdate: Any?
    private var syncTimer: Timer?
    private var backgrounded: Bool = true
```
nit ^

nbhasin2 commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,635 @@
+    fileprivate var backgrounded: Bool = true
+    private let logger: Logger
+
+    let FifteenMinutes = TimeInterval(60 * 15)
```
nit: 
```suggestion
    let fifteenMinutesInterval = TimeInterval(60 * 15)
```

nbhasin2 commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,635 @@
+    let FifteenMinutes = TimeInterval(60 * 15)
+
+    deinit {
+        if let c = constellationStateUpdate {
+            NotificationCenter.default.removeObserver(c)
+        }
```
nit: avoid single letter var
```suggestion
        if let stateUpdate = constellationStateUpdate {
            NotificationCenter.default.removeObserver(stateUpdate)
        }
```

tarikeshaq reviewed (changes requested):
Had a very brief look, will try to provide more insightful feedback early next week!

This is coming together 🎉

lmarceau commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,628 @@
+
+        let center = NotificationCenter.default
+
+        center.addObserver(self,
+                           selector: #selector(onStartSyncing),
+                           name: .ProfileDidStartSyncing,
+                           object: nil)
+        center.addObserver(self,
+                           selector: #selector(onFinishSyncing),
+                           name: .ProfileDidFinishSyncing,
+                           object: nil)
```
Two questions related to notifications:
1. So, we send those `.ProfileDidStartSyncing` & `.ProfileDidStartSyncing` from this `RustSyncManager` class and currently from BrowserProfile code too. Will we keep sending those notifications from BrowserProfile in the near future? 
2. If we send those notifications from `RustSyncManager`, is there added benefits of listening to the notifications `RustSyncManager`? Could we directly call `onStartSyncing` and `onFinishSyncing` instead?

lougeniaC64 replied:
1. So they're currently sent from `BrowserSyncManger` in Profile.swift. And the idea is that the `RustSyncManger` class that I'm adding in this PR will replace `BrowserSyncManager` during the Nimbus sync manager integration rollout. So there will be a lot of duplicated logic in both classes, but only one will ever be in use.
2. I suppose `RustSyncManager` doesn't need to listen to those notifications but they still need to be sent because there are a number of places in the Client code where those notifications are being observed. So, for instance, the `beginSyncing` function could call `onStartSyncing` but it would still need to emit the notification. This is more borrowed code so I don't have strong opinions here.

lmarceau commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,628 @@
+        // We just created the account, the user went through the Choose What to Sync
+        // screen on FxA.
+        if let declined = UserDefaults.standard.stringArray(
+            forKey: "fxa.cwts.declinedSyncEngines") {
+            declined.forEach { engineEnablements[$0] = false }
+            UserDefaults.standard.removeObject(forKey: "fxa.cwts.declinedSyncEngines")
```
nit, Looks like this key is used in other places, we could maybe have it as a constant?

lmarceau commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,628 @@
+        }
+
+        if !engineEnablements.isEmpty {
+            logger.log("""
+                engines to enable:
+                \(engineEnablements.compactMap { $0.value ? $0.key : nil })
+                """,
```
Wondering if this string format will save the log on multiple lines (I've had that issue in the past for other loggers), can we double check to make sure it's saved on one line?

lmarceau replied:
I had a look, and turns out yes using `"""` will make the log multiline in our logger too. This could make it harder to understand logs. I would suggest just using standard strings instead with `"`

My little test:
![Screen Shot 2023-03-06 at 5 38 38 PM](https://user-images.githubusercontent.com/11338480/223267219-0ebad250-65d6-482b-9406-f521fd389220.png)

![Screen Shot 2023-03-06 at 5 37 34 PM](https://user-images.githubusercontent.com/11338480/223267226-0f40fd02-ddf9-494c-86cb-148c8a71110c.png)

lougeniaC64 replied:
Sounds good! The only reason I used this format was to prevent the lines from being too long. 😊

lmarceau replied:
Yeah it's always a battle with strings to try to make them not too long haha. My only way otherwise to make them shorter is using variables like
```
let engines = "\(engineEnablements.compactMap { $0.value ? $0.key : nil })"
logger.log("engines to enable: \(engines)",
[...]
```

lmarceau commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,628 @@
+
+    fileprivate func getEnginesAndKeys(
+        engines: [String]
+    ) -> Deferred<Maybe<([EngineIdentifier], [String: String])>> {
```
Looking at the code, I guess it would be too much work to remove Deferred usage? 🥲

lougeniaC64 replied:
Yeah, because of the `SyncManager` protocol removing Deferred now would require a larger refactor. However, once the Nimbus sync manager integration rollout is over we can update the Deferred usage and refactor the callers.

lmarceau commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,628 @@
+
+                // XXX: `Scratchpad.clearFromPrefs` and `clearAll` were pulled from
+                // `SyncStateMachine.clearStateFromPrefs` to reduce RustSyncManager's
+                // dependence on the swift sync state machine logid. This will make
```
nit
```suggestion
                // dependence on the swift sync state machine logic. This will make
```

lmarceau commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,628 @@
+                                    .setString(syncResult.persistedState,
+                                               forKey: PrefsKeys.RustSyncManagerPersistedState)
+
+                                self.logger.log("""
+                                    Finished syncing with \(syncResult.status) status
+                                    """,
+                                    level: .info,
+                                    category: .sync)
+                                self.logger.log("""
+                                    Declined engines
+                                    \(String(describing: syncResult.declined ?? []))
+                                    """,
+                                    level: .info,
+                                    category: .sync)
+                                self.logger.log("""
+                                    Returned telemetry:
+                                    \(String(describing: syncResult.telemetryJson ??
+                                             "(No telemetry data was returned)"))
+                                    """,
+                                    level: .info,
+                                    category: .sync)
```
I think this can be done in one log only using the `extra` parameters on logger. We send info log as breadcrumbs to Sentry, so would be better to have this info noted in the same breadcrumb I think.

lmarceau commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,628 @@
+    }
+
+    public func syncHistory() -> OldSyncResult {
+        // XXX: The retrurn type of this function has been changed to comply with the
```
nit
```suggestion
        // XXX: The return type of this function has been changed to comply with the
```

lmarceau reviewed (commented):
Hello! I wanted to get familiar with the code too, I have some minor questions and nits. Let me know if there's anything. Thank you 🙏

lougeniaC64 commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,628 @@
+    ) -> Deferred<Maybe<MozillaAppServices.SyncResult>> {
+        let deferred = Deferred<Maybe<MozillaAppServices.SyncResult>>()
+
+        syncQueue.async { [unowned self] in
```
Sorry, I meant to leave I comment here. I purposely left this dispatch queue usage to illustrate an alternative because it was something Nish and I discussed and I was wondering if seeing the code might help. For sure we should only be using dispatch queues *_either_* here or in the lower level `RustSyncManagerAPI` functions.

@nbhasin2 Which is better?

nbhasin2 replied:
Ah, so we shouldn't create custom queues unless we need it. This is the mantra I go by and is what Apple agrees too. Ref: https://developer.apple.com/documentation/dispatch/dispatchqueue

However, there can be needs like 
- Performing task in a thread (syncQueue) that only your app is aware of
- Suspending the thread (syncQueue)
- Put a label so you can associate code specifically to your thread (syncQueue)

If you are doing any of this then we can leave `syncQueue` in our code. 

**IMPT Note:** Leaving it the way it is will not cause any harm but we should at least change `com.mozilla.firefox.sync` which is the same as Profile label to  `com.mozilla.firefox.sync.rust` to debug issues related to RustSyncManager 

**Final Thoughts** Change the label and leave the queue the way it is and we can come back to it for further review on how to leverage `syncQueue` and if we can't find anything in future then we use the Global Dispatch Queue

lougeniaC64 commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,628 @@
+
+    @discardableResult public func syncEverything(why: OldSyncReason) -> Success {
+        if let accountManager = RustFirefoxAccounts.shared.accountManager.peek(),
+           accountManager.accountMigrationInFlight() {
```
🫠 I'll remove that--thanks!

lougeniaC64 commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,628 @@
+
+        var stateHasBeenSet = false
+
+        // XXX: `RustTogglableEngines` is not used here because we want to sync the
```
Noting something that needs to be fixed later, which is true for all of the other instances I've used it except here. I'm just going to remove XXX so it doesn't cause confusion.

tarikeshaq commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,628 @@
+    }
+
+    func syncEverythingSoon() {
+        self.doInBackgroundAfter(SyncConstants.SyncOnForegroundAfterMillis) {
```
if I'm understanding correctly, the backgrounding of this is purely for the purpose of the delay correct? Like the actual syncing will dispatch back to the `syncQueue` (which in turn dispatches back to the `Dispatch.global` default queue?

lougeniaC64 replied:
I only intend to use either `syncQueue` or `Dispatch.global` depending on NIsh's preference.

tarikeshaq commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,628 @@
+        }
+    }
+
+    fileprivate func endRustSyncing(_ result: MozillaAppServices.SyncResult) {
```
naming nit: how about `endSyncing`? I see we use `startSyncing` etc

tarikeshaq commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,628 @@
+
+        syncDisplayState = resolveSyncState(result: result)
+
+        #if MOZ_TARGET_CLIENT
```
Do we need this and the `canSendUsageData`? I thought that with glean, telemetry settings are global (i.e glean won't send anything at all if disabled)?

lougeniaC64 replied:
Bet! I'll remove this.

tarikeshaq commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,628 @@
+                                    level: .warning,
+                                    category: .sync)
+                }
+            case "bookmarks":
+                if !registeredPlaces {
+                    self.profile.places.registerWithSyncManager()
+                    registeredPlaces = true
+                }
+                rustEngines.append("bookmarks")
+            case "history":
+                if !registeredPlaces {
+                    self.profile.places.registerWithSyncManager()
+                    registeredPlaces = true
+                }
+                rustEngines.append("history")
```
nit:
```suggestion
            case "bookmarks", "history":
                if !registeredPlaces {
                    self.profile.places.registerWithSyncManager()
                    registeredPlaces = true
                }
                rustEngines.append(engine)
```

lougeniaC64 commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,628 @@
+
+    func doInBackgroundAfter(_ millis: Int64, _ block: @escaping () -> Void) {
+        let queue = DispatchQueue.global(qos: DispatchQoS.background.qosClass)
+        // Pretty ambiguous here. I'm thinking .now was DispatchTime.now() and not
+        // Date.now()
+        queue.asyncAfter(
+            deadline: DispatchTime.now() + DispatchTimeInterval.milliseconds(Int(millis)),
```
More code from `BrowserSyncManager`. This comment was [added 10 months ago](https://github.com/mozilla-mobile/firefox-ios/blame/a9e338a97059600a5f376e680a99c87c2d49abca/Providers/Profile.swift#L898) so I wasn't sure, if this was relevant for upcoming work. I'll just remove it.

lougeniaC64 commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,628 @@
+    // or in an extension.
+    // But it's possible that we'll finish a side-effect sync after we've ditched the
+    // profile as a whole, so we hold on to our Prefs, potentially for a little while
+    // longer. This is safe as a strong reference, because there's no cycle.
```
So this was defintely copied as is so I'd defer to @nbhasin2 on this one, but now that we've removed `unowned` is anything in the comment accurate? Because now we're holding on to prefs and profile which seems to negate the premise of the comment.

And to amplify Tarik's point about the implications, are we comfortable with the effects of removing `unowned` here?

nbhasin2 replied:
Is all you are trying to do is access `Profile`?

lougeniaC64 replied:
Yeah I want some of the member functions and properties of the given profile.

nbhasin2 replied:
When once you have that initialization ready in another PR we can double-check to make sure if this will have a strong reference cycle or not. A class alone won't have any issues.

lougeniaC64 commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,628 @@
+
+        super.init()
+
+        let center = NotificationCenter.default
```
I actually removed this statement entirely when I removed the notification listeners as the `center` property was no longer needed.

lougeniaC64 replied:
Sorry, I refactored that code and lost the thread on why we needed the protocol, I'll add it.

lmarceau replied:
Perfect! As for the protocol, indeed there's not `object` param on the protocol. If we ever need it we can always add it. Thank you!

lougeniaC64 commented:
> Did another pass and dropped my thoughts going throughout 🙏 Most of it is questions and nits - I do think that testing this will be important
> 
> If I remember correctly, we wanted to get QA to test the branch before it lands on main, correct? Do we want to add any other automated tests, or do the ones covering SyncManager cover all the useful cases? (the sync what you want settings, all the meta request, clients syncing, etc

I've created a QA ticket for testing sync and send tab. I'm not sure exactly when in the process QA testing will happen, but I've been communicating with that team and they are expecting these changes. @nbhasin2 When can all of the changes in the sync manager epic branch be QA'd? Not sure what your plans are for merging it.

In regards to automated testing, the sync integration tests should still apply here. I will add some unit tests for the new logic I've added. Since the sync manager component handles client syncing on all syncs implicitly, I don't think that needs to be tested here.

mhammond commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,574 @@
+    lazy var syncManagerAPI = RustSyncManagerAPI(logger: logger)
+
+    public func migrateSyncData() {
+        // The sync functions used by the old native sync manager did not return a
```
I think I probably mislead you here before but this comment isn't quite correct.

The persisted global state is just a [list of engines which have been declined](https://github.com/mozilla/application-services/blob/main/components/sync15/src/client/state.rs#L42-L62) - there's no per-engine state, so it will not cause anything to think there was a first-sync.

The main risk of not doing this here will be that the first "real" sync will supply [enabled_changes](https://github.com/mozilla/application-services/blob/main/components/sync_manager/src/types.rs#L14) if the user happens to have enabled/disabled an engine exactly as this upgrade was happening, and because we didn't accurately know what was "declined" before we might do the wrong thing. The fact we can be 100% sure this first sync doesn't supply that arg means we can't get confused.

However, there's actually a reasonable chance that not asking for *any* engines to sync might do the right thing here too - ie, if we can somehow tell this magic we want to sync `SyncEngineSelection::Some{ engines: vec![] }` I suspect it will do what we want.

However, ISTM there's still a bit of a risk here - what if this sync fails due to, say, a transient network error? (I can't see where it's actually called though!) We are then lumping edge-cases on edge-cases so the chances of users doing a wrong thing might be small enough we don't really care. Alternatively, we could almost certainly construct a valid `persistedState` from what we know the enabled/disabled state is already (it's just a json string), which is info we already have - then we could avoid this entirely. (But is even *that* true - for all we know, the user tried to change the enabled state just before we upgraded, but that sync failed - so we can't be 100% sure the state of the UI checkboxes are an accurate representation of the current server state can we?

I wonder if another option is to simply not pass that `enabled_changes` param on a first sync based on a pref or something? It looks like that would mean this can die entirely too, and will still do the right thing after that first sync? I think that might work except for the caveat about that sync failing?

tl;dr - this needs more thought. Maybe we should start by working out exactly what might happen if that edge-case of the user changing a checkbox before the first normal sync happens and whether we are fine with that edge-case given it seems quite unlikely

lougeniaC64 replied:
Thanks for the clarification. So stepping back a bit, and I might be missing something here, but as the comment suggests I thought the reason for performing this "migration" was to prevent the first sync with the sync manager component taking longer than normal. If the actual issue is the edge case of the user changing their local engine enablements between the time we retrieve them and when they're passed into the sync call, is this an issue that's specific to the sync manager integration? In other words, is this an issue that could happen now in Firefox iOS release (or any other platform)?

I think for now I'm going to remove this function so that resolving how we handle this situation doesn't block the whole effort. In the meantime, I'll see if I can replicate this locally. Let me know if you have any objections @mhammond.

mhammond commented on Shared/Prefs.swift:
```diff
@@ -9,6 +9,13 @@ public struct PrefsKeys {
     public static let KeyLastRemoteTabSyncTime = "lastRemoteTabSyncTime"
+
+    // Global sync state for rust sync manager
+    public static let RustSyncManagerPersistedState = "rustSyncManagerPersistedStatKey"
```
Is `StatKey` a typo?

mhammond commented on Shared/Prefs.swift:
```diff
@@ -9,6 +9,13 @@ public struct PrefsKeys {
+    public static let RustSyncManagerPersistedState = "rustSyncManagerPersistedStatKey"
+
+    // Persisted State for rust sync manager has been migrated
+    public static let MigratedRustSyncPersistedState = "migratedRustSyncPersistedStateKey"
```
what's this one used for?

lougeniaC64 replied:
It's a flag for whether `migrateSyncData` has been  run. It's now obsolete since I'm removing that function, but basically `migrateSyncData` wouldn't be executed if  `MigratedRustSyncPersistedState` was true.

mhammond reviewed (commented):
LGTM, just still a bit confused about that migration stuff which we need to think through a bit more.

nbhasin2 commented:
@lougeniaC64 I can go in circles to improve the old code however as we discussed offline let's make the following changes for this PR 

a) make sure profile is marked `weak` and we can try to see if that causes any issues. Ideally, we wouldn't mark it `weak` but replace it with a delegate. Something like below but for now lets go with `weak` and come back to it to have the changes so we don't even have profile. This is to avoid `RustSyncManager` having direct access to `profile`
<img width="501" alt="image" src="https://user-images.githubusercontent.com/8919439/223843132-d6278b2c-256f-44e5-a472-37e4575cc31d.png">

b) Update queue names as those should be different for better debugging and if you don't need it you can simply use the global dispatch queue. 

Note: I believe you have already made these changes so I will one last walkthrough of this PR once the conflicts are resolved

mhammond commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,553 @@
+
+                            // Save the persisted state
+                            self.prefs
+                                .setString(syncResult.persistedState,
```
It's a bit of a bug in the rust code - `syncResult.persistedState` should probably be nullable to make the intent clearer, but I think you should persist this value any time it is not empty, even if the sync "failed"

mhammond commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,553 @@
+                                    }
+                                })
+                            } else {
+                                RustTogglableEngines.forEach({
```
This branch should be removed - it is nullable, and `null` will just mean we failed before we could determine what the error is. The branch above (ie, where you iterate over `declined` if it is non-null) should be done even if the success value isn't ok.

mhammond commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,553 @@
+                                           forKey: PrefsKeys.RustSyncManagerPersistedState)
+
+                            let declinedEngines = String(describing: syncResult.declined ?? [])
+                            let telemetryData = syncResult.telemetryJson ??
```
ditto here - we want to report failures too.

lougeniaC64 replied:
That `telemetryData` variable is for the logging. The telemetry is reported in the `endSyncing` function. Worth noting that the temp telemetry doesn't report status, which is just another reason for it to go.

mhammond reviewed (approved):
Thanks - this looks good although I think it needs a few minor tweaks for the sync failing scenario.

I'll try and see if my iOS build errors magically went away and give it a test, but if you can verify the declined engine stuff appears to work ok for you then I'm fine with landing it if you are 🚀

nbhasin2 commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,552 @@
+        profile?.pollCommands()
+    }
+
+    fileprivate func repeatingTimerAtInterval(
+        _ interval: TimeInterval,
+        selector: Selector
+    ) -> Timer {
```
```suggestion
    private func repeatingTimerAtInterval(_ interval: TimeInterval,
                                              selector: Selector) -> Timer {
```

nbhasin2 commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,552 @@
+        backgrounded = true
+    }
+
+    fileprivate func beginSyncing() {
```
nit: swift update
```suggestion
    private func beginSyncing() {
```

nbhasin2 commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,552 @@
+        notifySyncing(notification: .ProfileDidStartSyncing)
+    }
+
+    private func resolveSyncState(
+        result: MozillaAppServices.SyncResult
+    ) -> SyncDisplayState {
```
nit: indent
```suggestion
    private func resolveSyncState(result: MozillaAppServices.SyncResult) -> SyncDisplayState {
```

nbhasin2 commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,552 @@
+        }
+    }
+
+    fileprivate func endSyncing(_ result: MozillaAppServices.SyncResult) {
```
nit: swift update
```suggestion
    private func endSyncing(_ result: MozillaAppServices.SyncResult) {
```

nbhasin2 commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,552 @@
+        notificationCenter.post(name: notification)
+    }
+
+    init(profile: BrowserProfile,
```
nit: usually inits are at the beginning but in this file its way in the end, would be nice if you can move it more up

nbhasin2 commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,552 @@
+            execute: block)
+    }
+
+    var prefsForSync: Prefs {
```
nit: could we move this var near the top where other variables are initialized

nbhasin2 commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,552 @@
+        public let description = "Failed to get sync engine and key data."
+    }
+
+    fileprivate func getEnginesAndKeys(
+        engines: [String]
+    ) -> Deferred<Maybe<([EngineIdentifier], [String: String])>> {
```
```suggestion
    private func getEnginesAndKeys(engines: [String]) -> Deferred<Maybe<([EngineIdentifier],
                                                                             [String: String])>> {
```

nbhasin2 commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,552 @@
+        return deferred
+    }
+
+    fileprivate func syncRustEngines(
+        why: MozillaAppServices.SyncReason,
+        engines: [String]
+    ) -> Deferred<Maybe<MozillaAppServices.SyncResult>> {
```
nit: indent 

```suggestion
    private func syncRustEngines(why: MozillaAppServices.SyncReason,
                                 engines: [String]) -> Deferred<Maybe<MozillaAppServices.SyncResult>> {
```

nbhasin2 commented on Providers/RustSyncManager.swift:
```diff
@@ -0,0 +1,552 @@
+        return deferred
+    }
+
+    fileprivate func toSyncManagerDeviceType(
+        deviceType: DeviceType
+    ) -> SyncManagerDeviceType {
```
```suggestion
    private func toSyncManagerDeviceType(deviceType: DeviceType) -> SyncManagerDeviceType {
```

nbhasin2 commented:
@lougeniaC64 I see that at times you are using `MozillaAppServices.SyncResult` which could make the line longer. Below is a small suggestion to add it top the top of file and use it below. Let me know what you think?

<img width="1131" alt="image" src="https://user-images.githubusercontent.com/8919439/224179458-9beb602b-289a-4f7d-8013-8caabd68c547.png">

typealias: `private typealias MZSyncResult = MozillaAppServices.SyncResult`

Before
```
private func resolveSyncState(result: MozillaAppServices.SyncResult) -> SyncDisplayState {
```

After
```
private func resolveSyncState(result: MZSyncResult) -> SyncDisplayState {
```

nbhasin2 reviewed (approved):
Giving a preliminary approval as others are nits but we should revisit this in future to take a look at other areas we can improve. 

@lougeniaC64 Thanks for addressing so many comments, you are a trooper 🔥
:shipit:

nbhasin2 commented:
@lougeniaC64 thanks for creating the ticket, I am going to merge this into the epic branch
